AI Assistant: Track model used in toolbar extension and AI excerpt features

vianasw opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->
We need to track when we're using our default AI model and when we're using Chrome's AI built-in model. We also need to update the code to adapt to Chrome AI API changes.

Fixes JETPACK-326

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
*  Added two new events to track toolbar extension generate and undo actions (similar to events that we already have but these will have the model used)
* Added a model property in the AI excerpt tracking events
* Modified the code to adapt to Chrome AI API changes in the languageDetector and Translator API.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

Prerequisites:
* You'll need to test it using Chrome version 133 or higher.
* First you'll need a token to use the Chrome AI API. Check the comments on the issue JETPACK-326 for the tokens you need to use in the following snippet.
* Check out this branch, add the following snippet to enable beta features, the feature flag, and add your token in place of [ your trial token goes here ]:
```php
define( 'JETPACK_BLOCKS_VARIATION', 'beta' );
add_filter( 'ai_chrome_ai_enabled', '__return_true' );

$token_transient_name = 'jetpack-ai-chrome-ai-summarization-token';

set_transient(
    $token_transient_name,
    "[ your translation trial token goes here ]",
    3600
);


$token_transient_name = 'jetpack-ai-chrome-ai-translation-token';
set_transient(
    $token_transient_name,
    "[ your translation trial token goes here ]",
    3600
);

$token_transient_name = 'jetpack-ai-chrome-ai-language-detection-token';
set_transient(
    $token_transient_name,
    "[ your translation trial token goes here ]",
    3600
);
```
* Once this is set up, you should be able to create an excerpt using Chrome's AI API.
* Open the javascript console and enable debug mode.
* Create a new post and add some text.
* Go to the Excerpt section in the post sidebar and click on generate.
* The first time you click on it, the browser needs to download the model, which can take several seconds (close to a minute for me). There are no UI progress indicators for this, so you'll just have to be patient the first time. The following requests are almost instantaneous.
* Tone, word count, and language settings should work.
* The summarization API only returns summaries in English, check that if the content is in a different language, we fallback to using our default AI model.
* For the excerpt feature check that the following events have the `model` property and is `gemini-nano` or `default` for the fallback case:
  * jetpack_ai_assistant_block_generate
  * jetpack_ai_assistant_block_accept
  * jetpack_ai_assistant_block_discard
* Follow the instructions on https://github.com/Automattic/jetpack/pull/41724 to test the translation feature, but ignore the token setup part.
* For the translation feature check for these events: 
  * jetpack_ai_assistant_toolbar_extension_generate
  * jetpack_ai_assistant_toolbar_extension_undo

vianasw committed:
Add new AI model type constants

Added a default type and gemini-nano

vianasw committed:
Track AI model used in the AI Excerpt feature

We want to track when we're using Chrome AI bult-in AI model.

vianasw committed:
Adapt to latest Chrome AI API changes

The Translator and the languageDetector API had significant changes to
its API that I had to adapt.

vianasw committed:
Add tracking event for AI toolbar prompts

I've added two new events in order to track the acceptance rate of the
AI extension toolbar prompts (translate, change tone, etc):
- jetpack_ai_assistant_toolbar_extension_generate
- jetpack_ai_assistant_toolbar_extension_undo

I decided to add this two new events because it was easier to track the
AI model used in the requests (instead of having to pass the model all
the way down to where similar events are being tracked).

vianasw committed:
changelog

vianasw committed:
Add safety checks/fallbacks for Chrome summarizer

I've added safety checks to check if the summarizer API is available and
to check if the source language is English and I fall back to the
default model otherwise.

vianasw committed:
Adjust summarizer API changes

dhasilva commented on projects/js-packages/ai-client/src/chrome-ai/factory.ts:
```diff
@@ -123,14 +124,39 @@ export default async function ChromeAIFactory( promptArg: PromptProp ) {
+			return false;
+		}
+
+		if ( context.language ) {
```
This outer condition is not necessary, the inner condition is enough.

vianasw replied:
Yep, this is the result of rewrite number 4 or 5. I'll change it.

dhasilva commented on projects/js-packages/ai-client/src/chrome-ai/suggestions.ts:
```diff
@@ -165,15 +166,17 @@ export default class ChromeAISuggestionsEventSource extends EventTarget {
 
-		if ( available === 'after-download' ) {
+		let summarizer;
+		if ( available === 'available' ) {
+			summarizer = await self.ai.summarizer.create( summarizerOptions );
+		} else {
+			summarizer = await self.ai.summarizer.create( summarizerOptions );
 			await summarizer.ready;
 		}
```
This can be simplified

```suggestion
		summarizer = await self.ai.summarizer.create( summarizerOptions );

		if ( available !== 'available' ) {
			await summarizer.ready;
		} 
```

vianasw committed:
Update summarizer creation logic

Just simplify the summarizer creation logic

Co-authored-by: Douglas Henri <dhasilva@users.noreply.github.com>

vianasw committed:
Remove unnecessary if statement

vianasw committed:
Fix linter issue

vianasw committed:
Adapt to LanguageDetector API changes

LanguagetDetector API also changed and it was failing specifically with
the Arc browser (it uses the Chromium engine but it looks like it's not
100% compatible). I'm also taking this pass to make the implementation
more robust and fall back to the default model at any early sign that it
might fail.

vianasw committed:
Fix LanguageDetection availability check

CGastrell commented on projects/js-packages/ai-client/src/chrome-ai/factory.ts:
```diff
@@ -75,13 +75,43 @@ export default async function ChromeAIFactory( promptArg: PromptProp ) {
+	// If the languageDetector is not available, we can't use the translation or summary features—it's safer to fall back
+	// to the default AI model than to risk an unexpected error.
+	if (
+		! ( 'ai' in self ) ||
```
This check is failing, while I get `ai` object, `languageDetector` is not there. What I could dig about it is the [LanguageDetector API](https://developer.chrome.com/docs/ai/language-detection) might have changed its implementation. Once I enable the experimental flag, I get LanguageDetector's create and availability.

vianasw committed:
Adapt once more to API changes

Namespace changed again, I hope that for the last time in a while

CGastrell reviewed (approved):
Looking good! Worked as described on all scenarios.
